### PR TITLE
task policy

### PR DESCRIPTION
**Pull request by kakj-go**
#### What type of this PR

/kind feature

#### What this PR does / why we need it:
The pipeline task execution strategy is added. The new strategy can make the task no longer execute but directly fetch the latest successful result. This function can be used in automated testing, and users can save time to execute the same set of scenarios

#### Which issue(s) this PR fixes:

- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/all?id=228823&issueFilter__urlQuery=eyJzdGF0ZXMiOls0NDAyLDcxMDQsNzEwNSw0NDAzLDQ0MDQsNzEwNiw0NDA2LDQ0MDcsNDQxMiw0NTM4LDQ0MTMsNDQxNCw0NDE1LDQ0MTYsNDQwOF0sImFzc2lnbmVlSURzIjpbIjEwMDA1NjAiXSwiZmluaXNoZWRBdFN0YXJ0RW5kIjpbMTYzNjEyODAwMDAwMCxudWxsXX0%3D&issueTable__urlQuery=eyJwYWdlTm8iOjEsICJwYWdlU2l6ZSI6MTB9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=680&type=TASK)

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |      The task of the pipeline increases the configuration of the execution strategy        |
| 🇨🇳 中文    |       流水线的任务增加执行策略的配置       |

### CODE REVIEW

File: modules/pipeline/pipengine/reconciler/policy.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/erda-project/erda/modules/pipeline/dbclient"
+	"github.com/erda-project/erda/modules/pipeline/spec"
+	"github.com/erda-project/erda/pkg/parser/pipelineyml"
+)
+
+type PolicyHandlerOptions struct {
+	dbClient *dbclient.Client
+}
+
+var policyHandlers = map[apistructs.PolicyName]func(*spec.PipelineTask, PolicyHandlerOptions) (*spec.PipelineTask, error){
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/policy.go:30]: abstract a interface for handler.

File: modules/pipeline/pipengine/reconciler/policy.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	apistructs.LastSuccessResultPolicyType:    handlerLastSuccessResultPolicyType,
+	apistructs.FactoryNewPolicyType:           handlerFactoryNewPolicyType,
+}
+
+func handlerLastResultPolicyType(task *spec.PipelineTask, opt PolicyHandlerOptions) (result *spec.PipelineTask, err error) {
+	return task, nil
+}
+
+func handlerInPipelineLastResultPolicyType(task *spec.PipelineTask, opt PolicyHandlerOptions) (result *spec.PipelineTask, err error) {
+	return task, nil
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/policy.go:42]: Add TODO

File: modules/pipeline/pipengine/reconciler/policy.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	if task.Extra.Action.SnippetConfig == nil {
+		return task, nil
+	}
+
+	ymlName, source := getPipelineSourceAndNameBySnippetConfig(task.Extra.Action.SnippetConfig)
+
+	runSuccessPipeline, _, _, _, err := opt.dbClient.PageListPipelines(apistructs.PipelinePageListRequest{
+		Sources:        []apistructs.PipelineSource{source},
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/policy.go:57]: Do you need add snippet-config labels to query?

Comment (sfwn) [reply modules/pipeline/pipengine/reconciler/policy.go:57]: And I think this method is too huge for this situation.
Maybe you should add a new db method.

File: modules/pipeline/pipengine/reconciler/policy.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+		if dbTask.Type == task.Type && dbTask.Name == task.Name && dbTask.Status == apistructs.PipelineStatusSuccess {
+			beforeSuccessTask = &dbTask
+			break
+		}
+	}
+	if beforeSuccessTask == nil {
+		return task, nil
+	}
+
+	task.Status = beforeSuccessTask.Status
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/policy.go:89]: You should add a flag to distinguish this task's policy detail.

File: modules/pipeline/pipengine/reconciler/policy.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	})
+	if err != nil {
+		return task, err
+	}
+	if len(runSuccessPipeline) <= 0 {
+		return task, nil
+	}
+	pipeline := runSuccessPipeline[0]
+
+	tasks, err := opt.dbClient.ListPipelineTasksBySnippetPipelineID(&[]uint64{pipeline.ID}[0])
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/policy.go:73]: In logic, one snippet pipeline id should only bind to a certain parent task.

Comment (sfwn) [reply modules/pipeline/pipengine/reconciler/policy.go:73]: you should get parent task id from `pipeline_bases`.`parent_task_id`.

File: modules/pipeline/dbclient/op_pipeline_task.go [hunk 6]
```diff
@@ -146,6 +146,17 @@ func (client *Client) ListPipelineTasksByPipelineID(pipelineID uint64, ops ...Se
 	return tasks, nil
 }
 
+func (client *Client) ListPipelineTasksBySnippetPipelineID(snippetPipelineID *uint64, ops ...SessionOption) ([]spec.PipelineTask, error) {
```
Comment (sfwn) [inline modules/pipeline/dbclient/op_pipeline_task.go:149]: No need this db function.

File: modules/pipeline/pipengine/reconciler/policy.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+func handlerFactoryNewPolicyType(task *spec.PipelineTask, opt PolicyHandlerOptions) (result *spec.PipelineTask, err error) {
+	return task, nil
+}
+
+func getPipelineSourceAndNameBySnippetConfig(snippetConfig *pipelineyml.SnippetConfig) (ymlName string, sources apistructs.PipelineSource) {
+	return snippetConfig.Name, apistructs.PipelineSource(snippetConfig.Source)
+}
+
+func (r *Reconciler) adaptorPolicy(task *spec.PipelineTask) (result *spec.PipelineTask, err error) {
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/policy.go:109]: use verb. `adaptPolicy`

**Review by sfwn — CHANGES_REQUESTED** [review]

File: apistructs/pipeline_yml.go [hunk 8]
```diff
@@ -106,6 +106,28 @@ type PipelineYmlAction struct {
 	Disable       bool                   `json:"disable,omitempty"`                                        // task is disable or enable
 	Loop          *PipelineTaskLoop      `json:"loop,omitempty"`                                           // 循环执行
 	SnippetStages *SnippetStages         `json:"snippetStages,omitempty"`                                  // snippetStages snippet 展开
+	Policy        *Policy                `json:"policy,omitempty"`                                         // action 执行策略
+}
+
+type PolicyName string
```
Comment (sfwn) [inline apistructs/pipeline_yml.go:112]: PolicyType

File: apistructs/pipeline_yml.go [hunk 9]
```diff
@@ -106,6 +106,28 @@ type PipelineYmlAction struct {
...
+//
+// try-latest-success-result
+// force-latest-success-result
+//
+// new-run (default, can omit)
+//
+// run-once-from-root-pipeline
+const (
+	NewRunPolicyType               PolicyName = "new-run"
+	TryLastSuccessResultPolicyType PolicyName = "try-latest-success-result"
```
Comment (sfwn) [inline apistructs/pipeline_yml.go:126]: TryLatest

File: modules/pipeline/dbclient/op_pipeline_task.go [hunk 10]
```diff
@@ -145,7 +145,6 @@ func (client *Client) ListPipelineTasksByPipelineID(pipelineID uint64, ops ...Se
 	}
 	return tasks, nil
 }
-
```
Comment (sfwn) [inline modules/pipeline/dbclient/op_pipeline_task.go:148]: no need

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/reconciler/policy.go [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+	ymlName, source := getPipelineSourceAndNameBySnippetConfig(task.Extra.Action.SnippetConfig)
+
+	runSuccessPipeline, _, _, _, err := opt.dbClient.PageListPipelines(apistructs.PipelinePageListRequest{
+		Sources:        []apistructs.PipelineSource{source},
+		YmlNames:       []string{ymlName},
+		Statuses:       []string{apistructs.PipelineStatusSuccess.String()},
+		PageNum:        1,
+		PageSize:       1,
+		IncludeSnippet: true,
+		DescCols:       []string{apistructs.PipelinePageListRequestIdColumn},
```
[the remark this answers is not in the document]
Comment (kakj-go) [reply modules/pipeline/pipengine/reconciler/policy.go:65]: The effect of order by id is the same, and the query speed is faster

**Review by sfwn — APPROVED** [review]

**Pull request merged**